src/util/TunnelIPv6Interface.cpp

```diff
@@ -103,24 +103,79 @@ TunnelIPv6Interface::on_link_state_changed(bool isUp, bool isRunning)
 	syslog(LOG_INFO, "TunnelIPv6Interface::on_link_state_changed() UP=%d RUNNING=%d", isUp, isRunning);
 	if (isRunning != mIsRunning || isUp != mIsUp) {
 		if (isRunning && !mIsRunning) {
-			std::set<struct in6_addr>::const_iterator iter;
-			for (iter = mAddresses.begin(); iter != mAddresses.end(); ++iter) {
-				IGNORE_RETURN_VALUE(netif_mgmt_add_ipv6_address(mNetifMgmtFD, mInterfaceName.c_str(), iter->s6_addr, 64));
+			std::map<struct in6_addr, Entry>::iterator iter;
+
+			for (iter = mUnicastAddresses.begin(); iter != mUnicastAddresses.end(); ++iter) {
+				if (iter->second.mState != Entry::kWaitingToAdd) {
+					continue;
+				}
+
+				syslog(LOG_INFO, "Adding address \"%s/%d\" to interface \"%s\"",
+				       in6_addr_to_string(iter->first).c_str(), iter->second.mPrefixLen,
+				       mInterfaceName.c_str());
+
+				IGNORE_RETURN_VALUE(netif_mgmt_add_ipv6_address(mNetifMgmtFD, mInterfaceName.c_str(),
+				                    iter->first.s6_addr, iter->second.mPrefixLen));
+				iter->second.mState = Entry::kWaitingForAddConfirm;
 			}
-		}
 
+			for (iter = mMulticastAddresses.begin(); iter != mMulticastAddresses.end();	++iter) {
+				if (iter->second.mState != Entry::kWaitingToAdd) {
+					continue;
+				}
+
+				syslog(LOG_INFO, "Joining multicast address \"%s\" on interface \"%s\".",
+				       in6_addr_to_string(iter->first).c_str(), mInterfaceName.c_str());
+
+				IGNORE_RETURN_VALUE(netif_mgmt_join_ipv6_multicast_address(mNetifMgmtFD, mInterfaceName.c_str(),
+				                    iter->first.s6_addr));
+			}
+			mMulticastAddresses.clear();
```

Comment:
So this is more like a queue of pending multicast addresses to add than the list of multicast addresses...?